File: src/160.intersection-of-two-linked-lists.py
#
# @lc app=leetcode id=160 lang=python3
#
# [160] Intersection of Two Linked Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        currA: Optional[ListNode] = headA
        currB: Optional[ListNode] = headB
        while currA != currB:
            currA = currA.next if currA != None else headB
            currB = currB.next if currB != None else headA
        return currA
# @lc code=end

File: src/160.intersection-of-two-linked-lists.ts
/*
 * @lc app=leetcode id=160 lang=typescript
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

// function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
//   if (headA === null || headB === null) return null
//   const hashMap = new Map<ListNode, number>()
//   let currA: ListNode | null = headA
//   let currB: ListNode | null = headB
//   while (currA) {
//     hashMap.set(currA, currA.val)
//     currA = currA.next
//   }
//   while (currB) {
//     if (hashMap.has(currB)) return currB
//     currB = currB.next
//   }
//   return null
// }

function getIntersectionNode(headA: ListNode, headB: ListNode): ListNode | null {
  let currA: ListNode | null = headA
  let currB: ListNode | null = headB
  while(currA !== currB) {
    currA = currA ? currA.next : headB
    currB = currB ? currB.next : headA
  }
  return currA
}

// @lc code=end
